[DEV-11497] Socket connection management

kpsroka opened the pull request:
Rewrites the `PhoenixSocket` class to simplify connection management and message processing.

While this PR improves socket handling in the Phoenix library, it is not guaranteed to resolve any particular problem. Its intention is to primarily make the socket flow easier to understand and debug.

The major code change is that `PhoenixSocket` has been split into several classes, mainly `SocketConnectionManager` and its supporting classes.

The SocketConnectionManager is responsible for keeping a connection alive - it has facilities to keep track of current WebSocket connection state, reconnect connection on failures, and provides communication interface to `PhoenixSocket`. The upstream (to `PhoenixSocket` class) interface consists of three callbacks supplied to its constructor:
- `onMessage(String)` for receiving messages from the WebSocket connection,
- `onError(Object, [StackTrace])` for communicating errors with either initialization or operation of the connection.
- `onStateChange(WebSocketConnectionState)` which communicates current state of the connection (initialized, ready, closing, and closed).

The downstream interface is the `addMessage(String)` method.

In addition there are several smaller classes working with SocketConnectionManager:
- `_WebSocketConnection` which is a thin wrapper around `WebSocketChannel`, managing its stream/sink interface.
- `_ConnectionCallbacks` which encapsulates the upstream callbacks, and performs several checks to prevent obsolete or invalid calls from propagating.
- `SocketConnectionAttempt` which identifies connection attempt, and allows a bit of control over reconnection delay.
- `ConnectionInitializationException`.

The `PhoenixSocket` class is now mostly responsible for messaging and keeping track of the channels. What changes in some event handling is that this class no longer attempts to control the underlying WebSocket connection in response to the socket's events. Instead, the internals of `SocketConnectionManager` are responsible for controlling connections and acting on errors and closes to keep the connection running.

This PR also removes `PhoenixRawSocket`, which is unused, and incompatible with the changes.

Tests are pending, but I've been running it successfully as Mac and iOS app in the past day.

roughike commented:
I read through the changes. Before addressing the tiny details, I'd like to talk about the big picture.

I agree with you that the WTFs/minute when reading the library's source code is a bigger number that it should be. It's supposed to closely follow [the reference implementation](https://github.com/phoenixframework/phoenix/tree/main/assets/js/phoenix), which at times, it does not do a very good job at. However, many of us are familiar with the twisted way the library works, as we've debugged and tried to understand it for the past few years as we've fixed issues.

If we merge your PR, and replace a lot of old code with your new code, we'd be throwing a lot of that knowledge away. For the near future, this PR might simplify things on the surface level by splitting things up, but at a cost of replacing "the devil we know" with a new saint that we don't know that well yet.

> @kpsroka: While this PR improves socket handling in the Phoenix library, it is **not guaranteed to resolve any particular problem**. Its intention is to primarily make the socket flow **easier to understand and debug**.

This being the case, can we make very minimal (or zero?) changes to the existing code while achieving the same goal?

For example:
* Increase logging without changing existing code.
* Add more documentation comments about certain flows. For example, about the heartbeats (they're sent every 30s, if we receive server response, we effectively reset the heartbeat timeout), connection state changes, etc. Might be a double-edged sword if the documentation is not kept up to date.
* Add inline code comments at places where code is hard to read. I was going to say the other option is to make minimal changes to make the code more readable in separate PRs, but maybe it's better to stay as close to the reference implementation as possible so that it's easy to compare the two.
* If there are clear problems we need to improve, make purpose-built PRs for those.
* Something else?

As this PR currently stands, it feels a bit risky to merge it, especially when it does not fix anything concrete. We also don't have that many tests to begin with, so more minimal changes with a clear goal would be easier to manage.

kpsroka commented:
> I read through the changes. Before addressing the tiny details, I'd like to talk about the big picture.
> 
> I agree with you that the WTFs/minute when reading the library's source code is a bigger number that it should be. It's supposed to closely follow [the reference implementation](https://github.com/phoenixframework/phoenix/tree/main/assets/js/phoenix), which at times, it does not do a very good job at. However, many of us are familiar with the twisted way the library works, as we've debugged and tried to understand it for the past few years as we've fixed issues.
> 
> If we merge your PR, and replace a lot of old code with your new code, we'd be throwing a lot of that knowledge away. For the near future, this PR might simplify things on the surface level by splitting things up, but at a cost of replacing "the devil we know" with a new saint that we don't know that well yet.

I think that while we might be familiar with the current code, it being hard to understand also makes us avoid changing it even if we suspect something is wrong with it - it is more pleasant to fix our own code instead. The other benefit of having it easier is that more people will be able to learn it, and detect errors. I understand that this relies on relative simplicity of the new solution, so I'm open to updating this PR to simplify it even more, but I disagree with the general sentiment of valuing the current knowledge over future use.

> > @kpsroka: While this PR improves socket handling in the Phoenix library, it is **not guaranteed to resolve any particular problem**. Its intention is to primarily make the socket flow **easier to understand and debug**.
> 
> This being the case, can we make very minimal (or zero?) changes to the existing code while achieving the same goal?
> 
> For example:
> 
>     * Increase logging without changing existing code.
> 
>     * Add more documentation comments about certain flows. For example, about the heartbeats (they're sent every 30s, if we receive server response, we effectively reset the heartbeat timeout), connection state changes, etc. Might be a double-edged sword if the documentation is not kept up to date.
> 
>     * Add inline code comments at places where code is hard to read. I was going to say the other option is to make minimal changes to make the code more readable in separate PRs, but maybe it's better to stay as close to the reference implementation as possible so that it's easy to compare the two.
> 
>     * If there are clear problems we need to improve, make purpose-built PRs for those.
> 
>     * Something else?
> 

I don't think these can fully replace the code change themselves, and the result with applying these changes to existing code without any structural change can make the end result less readable.

> As this PR currently stands, it feels a bit risky to merge it, especially when it does not fix anything concrete. We also don't have that many tests to begin with, so more minimal changes with a clear goal would be easier to manage.

I probably should add more context: the intention for this PR was to resolve the duplicate connection problem by changes to the socket layer. But as with other PRs that we merged to resolve this problem, I cannot guarantee that it's going to fix it. From tests on my devices I didn't notice this problem surfacing\*, but I'll introduce more tests for this, including probable causes.


I'm putting this PR into draft mode for the time being, I'll focus on some tests and additional changes to simplify this even further.

\* I saw an instance of multiple [rapid connection/disconnection though](https://superlist.grafana.net/explore?schemaVersion=1&panes=%7B%22hn4%22:%7B%22datasource%22:%22grafanacloud-logs%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22dev-onsen%5C%22,%20container%3D%5C%22onsen%5C%22%7D%20%7C%3D%20%601-44fe-aae9-3134979928ca:iOS:1.12.2%60%20%7C%3D%20%60WebSocket%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22grafanacloud-logs%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%221723032204312%22,%22to%22:%221723032549831%22%7D,%22panelsState%22:%7B%22logs%22:%7B%22columns%22:%7B%220%22:%22Time%22,%221%22:%22Line%22%7D,%22visualisationType%22:%22logs%22,%22labelFieldName%22:%22labels%22%7D%7D%7D%7D&orgId=1). I'll investigate what's causing this [probably some calls to connect get queued too many times, when the only last one prevails]).

brian-superlist commented:
Summary of chat:

 - Naming and semantic meaning of `ready` and `connected`. Important for all of us to be on same page, would be good to hear from Iiro and Miguel about their opinion.
 - Want to ensure this version does not preclude us from using binary protobufs in the future. Didn't sound like this would create any blockers for that.

brian-superlist commented on lib/src/socket_connection_attempt.dart:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  SocketConnectionAttempt.aborted() {
+    _delayTimer = Timer(Duration.zero, () {});
```
This constructor is odd. Why does it set these variables up, then cancel them in the `abort` method? Why does it trigger an error in abort that is ignored here?

kpsroka replied:
Removed `.aborted()` in favor of using `null`.

brian-superlist commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,443 @@
+  /// If [immediately] is set to true, then an attempt to connect is made
+  /// immediately. This might result in dropping the current connection and
+  /// establishing a new one.
+  void start({bool immediately = false}) {
```
I think this should be called `connect` and `stop` should be called `halt`.

kpsroka replied:
By naming these methods like this I wanted to make it clearer that the "connection manager" is in indefinite started or stopped state, rather than performing a one-off action such as "connect". Does it make sense with this context?

brian-superlist commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,443 @@
+
+    // Make sure that no old attempt will begin emitting messages.
+    if (!_currentAttempt.delayDone) {
+      _currentAttempt.abort();
```
This aborted status is kinda weird as mentioned previously. Why not just null out `_currentAttempt`?

brian-superlist commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,443 @@
+        return;
+      }
+
+      _stopConnecting(4002, 'Immediate connection requested');
```
All status codes should have a semantic name. What is a `4002`?

kpsroka replied:
The codes in range [4000, 4999] are reserved for private use, so they don't have an inherent meaning for WebSocket. With 4001 and 4002 I wanted to signal to the backend special conditions that caused the closing of the connection. Right now we don't surface that on the backend, but it might allow us to track some close requests when we decide that we need them.

I will assign them to locally named constants.

brian-superlist commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,443 @@
+  /// success.
+  int _connectionAttempts = 0;
+
+  SocketConnectionAttempt _currentAttempt = SocketConnectionAttempt.aborted();
```
Why does it have an initial status of aborted? Shouldn't it be either `null` or `initial`? Semantically doesn't make much sense to initialize a variable in an `aborted` state when it hasn't been attempted.

kpsroka replied:
Ditto.

brian-superlist commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,443 @@
+            connectionCompleter
+                .completeError(StateError('Connection attempt aborted'));
+          } else {
+            // _startConnecting() was called during connection attempt, return the
```
What's `_startConnecting`?

kpsroka replied:
Old name for `_connect`. Updated.

kpsroka commented:
> Summary of chat:
> 
>  - Naming and semantic meaning of `ready` and `connected`. Important for all of us to be on same page, would be good to hear from Iiro and Miguel about their opinion.
>  - Want to ensure this version does not preclude us from using binary protobufs in the future. Didn't sound like this would create any blockers for that.

Replaced the PhoenixSocket._isConnected field with "_isOpen", which also is in line with the Phoenix state event naming.

Added some tests, and more comments. Will work on more tests tomorrow, but I think it's ready to review in detail now @roughike @miguelcmedeiros

brian-superlist commented on lib/src/socket.dart:
```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
-  }
+  Future<void> _startHeartbeat() async {
+    if (_socketStateStream.valueOrNull is! WebSocketReady) {
+      throw StateError('Cannot start heartbeat while disconnected');
```
I'd argue for renaming `WebSocketReady` -> `WebSocketConnected`. I understand they use `ready` under the hood for 1 method in the web socket channel package, but connected & disconnected are symmetrical. `ready`/ `disconnected` not as much.

kpsroka replied:
So would you also argue for renaming WebSocketClosing and WebSocketClosed to ...Disconnecting and ...Disconnected?

kpsroka replied:
Actually changed to Connecting, Connected, Disconnecting, Disconnected.

brian-superlist commented on lib/src/socket_state.dart:
```diff
@@ -0,0 +1,71 @@
+
+/// A connection attempt has started. This encompasses both local preparation of
+/// initial connection, and waiting for connection to become ready.
+final class WebSocketInitializing extends WebSocketConnectionState {
```
I preferred the `connecting` / `connected` language. Initializing doesn't necessarily mean connecting.

brian-superlist reviewed (commented):
Some inline comments around naming and also implementation.

brian-superlist commented on lib/src/socket.dart:
```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
+  }
+
+  bool _shouldPipeMessage(String message) {
+    final currentSocketState = _socketStateStream.valueOrNull;
```
Oh, forgot to leave a comment here. This could become a situation for a "black hole" where a user isn't receiving events but "we don't know why."

I can't say this will be a problem, but I wonder if we should throw an exception or log something here to sentry for the first version to ensure we aren't dropping anything we might want to keep.

kpsroka replied:
This is very much in line with how it used to work:
```dart
bool _shouldPipeMessage(dynamic event) {
  if (event is WebSocketChannelException) {
    return true;
  } else if (_socketState != SocketState.closed) {
    return true;
  } else {
    _logger.warning(
      'Message from socket dropped because PhoenixSocket is closed',
      '  $event',
    );
    return false;
  }
}
```

I agree that this is a bit dubious, and maybe we can change the `if` condition to allow also the disconnecting state.

brian-superlist commented on lib/src/socket.dart:
```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
+  // Exists for testing only
+  void onSocketDataCallback(String message) {
+    if (_shouldPipeMessage(message)) {
+      if (!_receiveStreamController.isClosed) {
```
Should this `!_receiveStreamController.isClosed` be inside `_shouldPipeMessage` as well?

kpsroka replied:
I find it clearer if it's with `_receiveStreamController.add(message);`. Let me know if you'd like me to change this.

brian-superlist replied:
👍 Not a strong opinion if you think it's clearer. Does it need the nested if statement though 🤔 

```
if (_shouldPipeMessage(message) && !_receiveStreamController.isClosed) {
        _receiveStreamController.add(message);
}
```

kpsroka replied:
Done.

brian-superlist commented on lib/src/socket.dart:
```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
+      if (message.ref != _latestHeartbeatRef) {
+        // The connection is alive, prevent heartbeat timeout from closing
+        // connection.
+        _latestHeartbeatRef = null;
```
Should we only clear this when the `message.ref != _latestHeartbeatRef` or should we just always clear the `_latestHeartbeatRef` when we receive any message as we do with the current implementation?

kpsroka replied:
Updated.

miguelcmedeiros commented on lib/src/socket.dart:
```diff
@@ -160,88 +158,38 @@ class PhoenixSocket {
+  /// Connects to the underlying WebSocket and prepares this PhoenixSocket for
+  /// connecting to channels.
+  ///
+  /// The returned future will complete as soon as the attempt to connect to a
```
We will have the see how the app code changes since the old `.connect()` completes when the connection is established.

miguelcmedeiros commented on lib/src/socket.dart:
```diff
@@ -290,46 +232,48 @@ class PhoenixSocket {
     _topicStreams.clear();
 
-    _stateStreamController.close();
+    _stateEventStreamController.close();
```
Close `_socketStateStream` as well?

kpsroka replied:
Done.

miguelcmedeiros commented on lib/src/socket.dart:
```diff
@@ -290,46 +232,48 @@ class PhoenixSocket {
+  /// a channel, you would usually use [PhoenixChannel.push] instead.
+  ///
+  /// Returns a future that completes when the reply for the sent message is
+  /// received. If your flow awaits for the result of this future, add a timout
```
```suggestion
  /// received. If your flow awaits for the result of this future, add a timeout
```

miguelcmedeiros commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,443 @@
+  /// If after call to [addMessage] a call to [dispose] or [stop] is made, then
+  /// this future will complete with an error instead.
+  Future<void> addMessage(String message) {
+    return _maybeConnect().then((connection) => connection.send(message));
```
```suggestion
    final connection = await _maybeConnect();
    return connection.send(message);
```

kpsroka replied:
Done.

miguelcmedeiros commented on lib/src/socket.dart:
```diff
@@ -338,22 +282,21 @@ class PhoenixSocket {
       );
     }
     _addToSink(_options.serializer.encode(message));
+    if (message.event != PhoenixChannelEvent.heartbeat) {
```
Remove this in favour of only cancelling when receiving a message.

kpsroka reviewed (commented):
> This constructor is odd. Why does it set these variables up, then cancel them in the `abort` method? Why does it trigger an error in abort that is ignored here?  

Replaced with something a bit more structured but less weird.

brian-superlist commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,444 @@
+      try {
+        connection = await _runConnectionAttempt(delay);
+      } catch (error, stackTrace) {
+        _logger.warning('Failed to initialize connection', error, stackTrace);
```
I'm a little worried about the flow here. This is a black hole. If something goes wrong, we simply swallow it and log it to nowhere.

This can hide things like invalid tokens, etc, since `_runConnectionAttempt` calls the `buildMountPoint` under the hood.

kpsroka replied:
Error reporting is done in `_runConnectionAttempt` (unless we obsolete it with another attempt).

kpsroka reviewed (commented):
Adressed most of the comments, PTAL.

Also renamed `PhoenixSocket.isConnected` to `isOpen`, not to confuse it with connection state.

kpsroka commented:
@brian-superlist @miguelcmedeiros @roughike 

Updated structurally how the code is split in various classes, with main intention being simplicity, and preservation of current behavior of PhoenixScoket.connect().

What changed:
- `SocketConnectionAttempt` got renamed to `DelayedCallback`, and added a callback to be executed when the delay completes. This is now a more generic class, effectively serving as `Future.delayed` with control over delay. I wanted to preserve it as a dependency-less utility, so it has no connection (nominal or logical) to "sockets".
- `SocketConnectionManager` is now ___unstoppable___ - I removed the "stop" facility, so effectively once started it can only be disposed. This allowed for simpler life-cycle of it. However, now it needs to be reinstantiated whenever `PhoenixSocket` closes (disposing of the manager) and reopens.

miguelcmedeiros commented on lib/src/socket.dart:
```diff
@@ -341,20 +311,16 @@ class PhoenixSocket {
-      return;
-    }
-    _ws!.sink.close(code, reason);
+    _connectionManager!.addMessage(data);
```
Should we await this?
And in `sendMessage()`, should we await `_addToSink`

kpsroka replied:
Refactored this a little now to allow for awaiting on `ConnectionManager.addMessage`, but not on completion of response future.

miguelcmedeiros commented on lib/src/socket_connection.dart:
```diff
@@ -0,0 +1,444 @@
+        immediately: true,
+      );
+    }
+    await _maybeConnect();
```
Is it safe to call `_maybeConnect()` if `immediately == true`?
`reconnect()` also calls `_connect()`

kpsroka replied:
It is safe, but just to be clear made it so that we return inside the `if`.

brian-superlist commented:
Hey all, I'm sorry to leave this comment right before I leave, but I'm not sure I fully support merging this PR. Right now, the network connection is pretty stable and recovers well. We have not had users writing in reports about seeing the offline indicator, nor are we sure we are still running into double connection issues. Finally, it looks like we are failing on a few different integration tests now.

Therefore, if this PR does not solve a specific issue, while possibly causing some regressions, I'm not sure if it's worth it. Readability is in the eye of the beholder, and I personally think this is harder to reason about than the previous code. Miguel mentioned that I might have Stockholm syndrome 😅, and I actually sympathize with that description. However, to make such a large change to a library that is now quite stable, we should have a strong case for it: What specifically does it improve?

I don't want to block the merging of this PR if the team feels it's the right direction, but I wanted to note that even after these latest changes I'm also hesitant about this changeset and the possibility that it might cause us new headaches with connections right after we solved so many of them.

If we want to continue with the plan we have right now: Fix up the integration tests, merge early next week, and test, I can disagree and commit. However, I wanted to try to clarify that I'm still not fully sold on this set of changes.

roughike commented:
Just for the record - I echo Brian's comment and what I said 2 weeks ago.

Feels like the risk-to-reward ratio is off with this one. I don't feel like it's the right direction, I don't think the changes simplify anything, and I think actually makes things more hard to follow for me.

I don't want to merge this one. If you and @miguelcmedeiros feel strongly about merging this one, I will not block you, but I'll assume that you know what you're doing and know the risks.

If you do decide to merge, at least the failing integration tests must be fixed or there has to be a very good reason why not to fix them.

MarcelKaeding commented:
Let's merge **not** before the 1.14.0 build was sent to Testlio. We can then leave it in internal testing for 1 cycle before we release it to users.

kpsroka commented:
> Just for the record - I echo Brian's comment and what I said 2 weeks ago.
> 
> Feels like the risk-to-reward ratio is off with this one. I don't feel like it's the right direction, I don't think the changes simplify anything, and I think actually makes things more hard to follow for me.

I can add additional tests over the ones that are here, and that I've added myself, to even further minimize the risk.

> I don't want to merge this one. If you and @miguelcmedeiros feel strongly about merging this one, I will not block you, but I'll assume that you know what you're doing and know the risks.

That's why I'd like to have it tested internally on daily before pushing it to production.

> If you do decide to merge, at least the failing integration tests must be fixed or there has to be a very good reason why not to fix them.

After a bit of digging I realized that the backend at `example/backend` is a backend for integration testing (maybe _the_ backend). I removed two tests which were not passing since (at least) 658661b0f673470ca9e8fb70eec7c19b02061db0, and adjusted some - mostly to call `socket.dispose()` either instead of `socket.close()` or actually running `socket.dispose()` at all.
